src/examples/libpmemobj/buffons_needle_problem.c

```diff
@@ -0,0 +1,161 @@
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * buffons_needle_problem.c <path> <n> or
+ *		buffons_needle_problem.c <path> -- example illustrating
+ *		usage of libpmemobj
+ *
+ * Calculates pi number by solving Buffon's needle problem.
+ * Takes one/two arguments -- path of the file and integer amount of trials
+ * or only path when continuing simulation after interruption.
+ * The greater number of trials, the higher calculation precision.
+ */
+
+#include <ex_common.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <limits.h>
+#include <math.h>
+#include <time.h>
+#include <libpmemobj.h>
+/*
+ * Layout definition
+ */
+POBJ_LAYOUT_BEGIN(pi);
+POBJ_LAYOUT_ROOT(pi, struct my_root)
+POBJ_LAYOUT_END(pi)
+/*
+ * Used for changing degrees into radians
+ */
+#define RADIAN_CALCULATE M_PI / 180.0
+
+static PMEMobjpool *pop;
+
+struct my_root {
+	double x; /* coordinate of the needle's center */
+	double angle; /* angle between vertical position and the needle */
+	double l; /* length of the needle */
+	double sin_angle_l; /* sin(angle) * l */
+	double pi; /* calculated pi number */
+	double d; /* distance between lines on the board */
+	uint64_t i; /* variable used in for loop */
+	uint64_t p; /* amount of the positive trials */
+	uint64_t n; /* amount of the trials */
+};
+
+static void
+print_usage(char *argv_main[])
+{
+	printf("usage: %s file-name n or %s file-name\n",
+		argv_main[0], argv_main[0]);
+}
+
+int
+main(int argc, char *argv[])
+{
+	if (argc < 2 || argc > 3) {
+		print_usage(argv);
+		return 1;
+	}
+
+	const char *path = argv[1];
+
+	if (file_exists(path) != 0) {
+		if ((pop = pmemobj_create(path, POBJ_LAYOUT_NAME(pi),
+				PMEMOBJ_MIN_POOL, 0666)) == NULL) {
+			perror("failed to create pool\n");
+			return 1;
+		}
+	} else {
+		if ((pop = pmemobj_open(path,
+				POBJ_LAYOUT_NAME(pi))) == NULL) {
+			perror("failed to open pool\n");
+			return 1;
+		}
+	}
+
+	srand(time(NULL));
+	TOID(struct my_root) root = POBJ_ROOT(pop, struct my_root);
+
+	if (argc == 3) {
+		const char *n = argv[2];
+
+		char *endptr;
+		errno = 0;
+		uint64_t ull_n = strtoull(n, &endptr, 10);
+
+		if (*endptr != '\0' ||
+				(ull_n == ULLONG_MAX && errno == ERANGE)) {
+			perror("Wrong n parameter\n");
+			print_usage(argv);
+			return 1;
+		}
+
+		TX_BEGIN(pop) {
+			TX_ADD(root);
```

Comment:
given how often you use the root variable, I think it would be just easier to make a local variable with the real pointer, like so:
`struct my_root *rootp = D_RW(root);` That's the recommended approach, especially in tight loops.